### PR TITLE
Unify JSON decoding between the direct and spooled protocols

### PR DESCRIPTION
**Pull request by wendigo**
This switches existing protocol JSON parsing to a new, JsonParser-based approach that is much faster and requires less memory.

## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```
## CLI/JDBC/Client
* Improve performance and memory consumption when decoding data.
```

### CODE REVIEW

File: client/trino-client/src/main/java/io/trino/client/JsonResultRows.java [hunk 1]
```diff
@@ -37,33 +36,15 @@
 import static com.fasterxml.jackson.core.JsonToken.START_ARRAY;
 import static com.google.common.base.Verify.verify;
 import static com.google.common.collect.Iterators.unmodifiableIterator;
+import static io.trino.client.JsonDecodingUtils.createTypeDecoders;
 import static java.util.Collections.unmodifiableList;
 import static java.util.Objects.requireNonNull;
 
-class JsonResultRows
-        implements ResultRows
+public class JsonResultRows
```
Comment (losipiuk) [inline client/trino-client/src/main/java/io/trino/client/JsonResultRows.java:43]: can you make class final?

File: client/trino-client/src/main/java/io/trino/client/JsonResultRows.java [hunk 2]
```diff
@@ -73,22 +54,27 @@ private static class RowWiseIterator
         private final JsonParser parser;
         private final TypeDecoder[] decoders;
 
-        public RowWiseIterator(InputStream stream, TypeDecoder[] decoders)
+        public RowWiseIterator(JsonParser parser, TypeDecoder[] decoders)
```
Comment (losipiuk) [inline client/trino-client/src/main/java/io/trino/client/JsonResultRows.java:57]: drop this `implements Iterable` abomination

File: client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonQueryDataDecoder.java [hunk 3]
```diff
@@ -14,15 +14,16 @@
...
 import io.trino.client.ResultRows;
 import io.trino.client.spooling.DataAttributes;
-import io.trino.client.spooling.encoding.JsonDecodingUtils.TypeDecoder;
 
 import java.io.InputStream;
 import java.util.List;
 
-import static io.trino.client.spooling.encoding.JsonDecodingUtils.createTypeDecoders;
+import static io.trino.client.JsonDecodingUtils.createTypeDecoders;
+import static io.trino.client.JsonResultRows.forInputStream;
```
Comment (losipiuk) [inline client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonQueryDataDecoder.java:26]: do not static import

File: client/trino-client/src/main/java/io/trino/client/ResultRowsDecoder.java [hunk 4]
```diff
@@ -32,7 +32,7 @@
 import static com.google.common.base.Preconditions.checkState;
 import static com.google.common.base.Verify.verify;
 import static com.google.common.collect.ImmutableList.toImmutableList;
-import static io.trino.client.FixJsonDataUtils.fixData;
+import static io.trino.client.JsonResultRows.forJsonParser;
```
Comment (losipiuk) [inline client/trino-client/src/main/java/io/trino/client/ResultRowsDecoder.java:35]: do not static imnport

**Review by losipiuk — APPROVED** [review]
A bit hard to follow but looks good

File: client/trino-client/src/main/java/io/trino/client/JsonResultRows.java (hunk 2 above)
Comment (wendigo) [reply client/trino-client/src/main/java/io/trino/client/JsonResultRows.java:57]: Good point!
